waiter/src/waiter/token.clj

```diff
@@ -719,3 +763,125 @@
+              (reset! state-atom current-state)
+              (let [[msg current-chan]
+                    (async/alts! [exit-chan tokens-update-chan tokens-watch-channels-update-chan
+                                  watch-refresh-timer-chan query-chan]
+                                 :priority true)
+                    next-state
+                    (condp = current-chan
+                      exit-chan
+                      (do
+                        (log/warn "Stopping token-watch-maintainer")
+                        (when (not= :exit msg)
+                          (throw (ex-info "Stopping router-state maintainer" {:time (t/now) :reason msg}))))
+
+                      tokens-update-chan
+                      (timers/start-stop-time!
+                        (metrics/waiter-timer "core" "token-watch-maintainer" "process-tokens-update")
+                        (let [{:keys [token owner]} msg
+                              token-index-entry (get-token-index kv-store token owner :refresh true)
+                              local-token-index-entry (get token->token-index token)]
+                          (cond
+                            ; There is no change detected, so no event to be reported
+                            (= token-index-entry local-token-index-entry)
+                            current-state
+                            ; If index-entry retrieved from kv-store exists then treat as UPDATE (includes token creation)
+                            (some? token-index-entry)
+                            (do
+                              (->> token-index-entry
+                                   (make-index-event :UPDATE)
+                                   (async/>! tokens-event-chan))
+                              (assoc-in current-state [:token->token-index token] token-index-entry))
+                            :else
+                            (do
+                              (->> {:owner owner :token token}
+                                   (make-index-event :DELETE)
+                                   (async/>! tokens-event-chan))
+                              (assoc current-state :token->token-index (dissoc token->token-index token))))))
+
+                      tokens-watch-channels-update-chan
+                      (timers/start-stop-time!
+                        (metrics/waiter-timer "core" "token-watch-maintainer" "process-tokens-channels-update")
+                        (let [{:keys [channel-event watch-chan]} msg
+                              next-state (case channel-event
+                                           :add (do
+                                                  (async/put! watch-chan
+                                                              (make-index-event :INITIAL
+                                                                                (or (vals token->token-index) '())))
```

Comment:
Please use an empty vector `[]` instead of a list. Seeing the use of list literal is unusual in the codebase.